types/payload.go

```diff
@@ -0,0 +1,235 @@
+package types
+
+import (
+	"encoding/json"
+	"iter"
+	"maps"
+
+	jsoniter "github.com/json-iterator/go"
+	"github.com/vmihailenco/msgpack/v5"
+)
+
+type Payload struct {
+	// A serialized messagepack map used to source fields.
+	msgpMap *MsgpPayloadMap
+
+	// Deserialized fields, either from the internal msgpMap, or set externally.
+	memoizedFields map[string]any
+}
+
+func NewPayload(data map[string]any) Payload {
+	return Payload{
+		memoizedFields: data,
+	}
+}
+
+func (p *Payload) UnmarshalMsgpack(data []byte) error {
+	p.msgpMap = &MsgpPayloadMap{rawData: data}
+	p.memoizedFields = make(map[string]any)
+	return nil
+}
+
+func (p *Payload) UnmarshalJSON(data []byte) error {
+	var fields map[string]any
+	if err := jsoniter.Unmarshal(data, &fields); err != nil {
+		return err
+	}
+	p.memoizedFields = fields
+	return nil
+}
+
+// Extracts all of the listed fields from the internal msgp buffer in a single
+// pass, for efficient random access later.
+func (p *Payload) MemoizeFields(keys ...string) {
+	if p.msgpMap == nil {
+		return
+	}
+
+	if p.memoizedFields == nil {
+		p.memoizedFields = make(map[string]any, len(keys))
+	}
+
+	keysToFind := make(map[string]struct{}, len(keys))
+	for _, key := range keys {
+		if _, ok := p.memoizedFields[key]; !ok {
+			keysToFind[key] = struct{}{}
+		}
+	}
+	if len(keysToFind) == 0 {
+		return
+	}
+
+	iter, err := p.msgpMap.Iterate()
+	if err != nil {
+		return
+	}
+
+	var keysFound int
+	for keysFound < len(keysToFind) {
+		keyBytes, _, err := iter.NextKey()
+		if err != nil {
+			break
+		}
+
+		// Note we deliberately don't put string(keyBytes) in a variable here,
+		// because doing so will move it to the heap on every iteration.
+		// Keeping the string cast inline like this allows us to avoid the heap
+		// unless we're actually going to memoize the field.
+		if _, ok := keysToFind[string(keyBytes)]; ok {
+			value, err := iter.ValueAny()
+			if err != nil {
+				break
+			}
+			key := string(keyBytes)
+			p.memoizedFields[key] = value
+			keysFound++
+		}
+	}
+}
+
+func (p *Payload) Get(key string) any {
+	if p.memoizedFields != nil {
+		if value, ok := p.memoizedFields[key]; ok {
+			return value
+		}
+	}
+
+	iter, err := p.msgpMap.Iterate()
+	if err != nil {
+		return nil
+	}
+
+	for {
+		keyBytes, _, err := iter.NextKey()
+		if err != nil {
+			break
+		}
+
+		if string(keyBytes) == key {
+			value, err := iter.ValueAny()
```

Comment:
What's the performance tradeoff here by just using ValueAny()? What about using the Value method for concrete types?